feat: embedder service

jordanh opened the pull request:
# Description

Cleaned up for company, this is the embedder service that is being readied for merge.

While there are a few enhancements that I'd like to make in short order, this PR intends only to target our local development environments. When you review, keep your eye on that rather than asking, "is this ready for production?"

Enhancements to make after this PR merges:

- Added OpenAI embedding and GPT API support (rather than local dev services)
- Resume events that are stuck in a queued state
- Made fully event-driven (rather than use a polling loop)

**👉 Consider started your review by reading the [README](https://github.com/ParabolInc/parabol/blob/e0746635ac38fa459da134a8a19cd870b731f087/packages/embedder/README.md), then beginning your code review [here](https://github.com/ParabolInc/parabol/pull/9417/files#r1484536083)**

## Demo

📺 https://www.loom.com/share/524994e87dc34c5db360d3045d576e62?sid=789e6b1e-9ef2-4d18-9f96-ee4476426753

## Testing scenarios

### Requirements

- [ ] Checkout and run https://github.com/ParabolInc/node-llama-cpp-text-generation-interface

- [ ] Update your `.env` from `.env.example`
   - [ ] Set: `POSTGRES_USE_PGVECTOR=true`
   - [ ] Set: `AI_EMBEDDER_ENABLED`
   - [ ] Set `AI_EMBEDDING_MODELS`
   - [ ] Set: `AI_GENERATION_MODELS`

- [ ] Add the `relatedDiscussions` feature flag to any org you'd like retro discussions to be embedded ([see here](https://github.com/ParabolInc/parabol/pull/9417/files#r1484542278))

### Tests

- [ ] Can run embedder service via `yarn dev`
  - Observes rows being created in `EmbeddingsMetadata`
  - Observes rows entering/leaving `EmbededingsJobQueue`
  - Observes rows being created into configured embedding vector table, e.g. `Embeddings_ember_1`

## Final checklist

- [ ] I checked the [code review guidelines](../docs/codeReview.md)
- [ ] I have added [Metrics Representative](../docs/codeReview.md#metrics-representative) as reviewer(s) if my PR invovles metrics/data/analytics related changes
- [ ] I have performed a self-review of my code, the same way I'd do it for any other team member
- [ ] I have tested all cases I listed in the testing scenarios and I haven't found any issues or regressions
- [ ] Whenever I took a non-obvious choice I added a comment explaining why I did it this way
- [ ] I added the label https://github.com/ParabolInc/parabol/labels/Skip%20Maintainer%20Review if the PR introduces only minor changes, does not contain any architectural changes or does not introduce any new patterns and I think one review is sufficient'
- [ ] PR title is human readable and could be used in changelog

jordanh commented on package.json:
```diff
@@ -103,8 +103,8 @@
     "jscodeshift": "^0.14.0",
-    "kysely": "^0.26.3",
-    "kysely-codegen": "^0.10.0",
+    "kysely": "^0.27.2",
```
While I was in the neighborhood, I bumped kysely

jordanh commented on packages/embedder/ai_models/TextGenerationInterface.ts:
```diff
@@ -0,0 +1,80 @@
+import {
```
A thought for the future:

In the future, we could have deployments with local LLMs. It might be nice for us to have a generic GPT interface that abstracts away whether or not we are using a local LLM or OpenAI, etc. and have the executor services consume that interface rather than call OpenAI directly

jordanh commented on packages/embedder/embedder.ts:
```diff
@@ -0,0 +1,250 @@
+          continue
+        }
+      }
+      // console.log(`embedText: ${embedText}`)
```
uncomment me if you want to see lots of fun user data

jordanh commented on packages/embedder/embedder.ts:
```diff
@@ -0,0 +1,250 @@
+
+const tick = async (modelManager: ModelManager) => {
+  console.log(`embedder: tick`)
+  await redlock
```
Using a mutex lock allows us to run multiple, parallel embedder processes in a stateless fashion. The critical section of the embedder process ensures that only one embedder process is scraping our DB at a time for historical objects. This is particularly handy for rethinkdb, as it has no native locking constructs.

An alternative would be to define another table in postgres where each row contained something like:

   - objectType
   - timeStart
   - timeStop

And each embedder process would grab a time range with a row lock and begin indexing.

In the end, I decided to achieve parallel safety using the redis mutext lock and a queue as it was simpler to implement than needing to manage Postgres table/row locking.

jordanh commented on packages/embedder/embedder.ts:
```diff
@@ -0,0 +1,250 @@
+  return envVarValue === 'true'
+}
+
+const run = async () => {
```
Dear reader, I recommend starting your journey here here in this run function

jordanh commented on packages/server/dataloader/customLoaderMakers.ts:
```diff
@@ -478,11 +478,11 @@ export const meetingTemplatesByOrgId = (parent: RootDataLoader) => {
         .where('orgId', 'in', orgIds)
         .where('isActive', '=', true)
-        .where(({or, cmpr}) =>
+        .where(({or, eb}) =>
```
updated when I bumped kysely; cmpr was deprecated, then removed

jordanh commented on packages/server/dataloader/customRedisQueries.ts:
```diff
@@ -38,11 +38,11 @@ const customRedisQueries = {
-          .where(({or, cmpr}) =>
+          .where(({or, eb}) =>
             or([
-              cmpr('hideStartingAt', 'is', null),
```
updated when I bumped kysely; cmpr was deprecated, then removed

jordanh commented on scripts/webpack/toolbox.config.js:
```diff
@@ -56,6 +56,26 @@ module.exports = {
     new webpack.DefinePlugin({
       __PRODUCTION__: true
     })
+    // new CircularDependencyPlugin({
```
This was helpful for debugging why I was having difficulty with a webpack build at one point. Decided to leave the full config in here if it helps anybody else in the future

mattkrick commented on packages/embedder/indexing/retrospectiveDiscussionTopic.ts:
```diff
@@ -0,0 +1,267 @@
+
+export async function refreshEmbeddingsMeta(dataLoader: DataLoaderWorker) {
+  const r = await getRethink()
+  const {createdAt: newestMeetingDate} = (await r
```
-1 if you use `createdAt` here, but filter by `summarySentAt` below, then meetings that haven't ended by the time the tick comes along will never get handled, unless i misread this?

jordanh replied:
Indeed you're correct. I'll add a comment that this process is merely for building history. The meetings missed by this function (summarySentAt is `null`) will get added to the meta when the meetings complete by adding a hook to the main app.

mattkrick replied:
Help me understand further. If we put a hook in the main app that adds a row to the metadata table, that would make `newestMetadata == Date.now()`, which would then be used as the `startDateTime` so no more history would get processed, right?

mattkrick replied:
whoops, forgot to un-resolve this issue. this is probably the biggest question mark i've got with the PR. we can merge it, but i wouldn't run the embedder until we get this one figured out otherwise it won't know where to start

jordanh replied:
@mattkrick Oh wow. Good call. I just got what you were inquiring about.

I can think of only one good option. In essence we need a high-water mark for historically items. I think it might look like this:

1. **Establish high-water mark**: we create a table called `EmbeddingsWatermark` or `EmbeddingsHistory` or something

```
id                  PRIMARY KEY
objectType          EmbeddingsObjectTypeEnum
newest              TIMESTAMP
```

2. **Process history using this watermark**: only the embedder process (and the one that has won the election at that) adjusts `newest` after adding each batch to `EmbeddingsMetadata`. If objects are added directly to Metadata directly, this field is not updated.

What do you think?

jordanh replied:
(I also think we should handle this as a separate PR, but let's get the issue added with clear criteria before merging)

mattkrick replied:
agreed, we need a watermark. if the operation takes more than a couple minutes, we could store that in a table. if the operation is quick, we can just tell it the date range to process. Either way, we need something that supports processing a particular date range in the case of a bug. Quick pseudo code 

```ts
// redis CLI
PUBLISH buildObjects "{start: '1970-01-01', end: '2024-02-22', objectType: 'discussionTopic'}"

// embedder.ts
const subscriber = new RedistInstance('sub')
subscriber.subscribe('buildObjects')
subscriber.on('message', _channel, message) => {
    addMetadata(JSON.parse(message))
}

const addMetadata = async ({start: Date, end?: Date, objectType?: ObjectType}) => {
  if (objectType === 'discussionTopic') {
    addMetaDataForDiscussions(start, end)
  }
}

const addMetadataForDiscussions(start: Date, end?: Date) {
  let curStart = start
  for (i = 0; i <1e6; i++) {
    const items = r.table('NewMeeting').between(curStart, end, {index: 'endedAt'}).limit(100)
    const failedToInsertDueToUniqueConstraintViolation = await processItems(items)
    if (items.length < 100 || failedToInsertDueToUniqueConstraintViolation) break
    curStart = items[items.length - 1].endedAt
  }
}
```

jordanh replied:
I like breaking up the function this way.

We're always going to need to keep the watermark of the newest historical range in storage, regardless; as you pointed we need  a way to create the time ranges when the embedder is chewing through history. I'll capture both of these things in an issue.

jordanh replied:
Added issue here: https://github.com/ParabolInc/parabol/issues/9451

mattkrick commented on packages/embedder/indexing/retrospectiveDiscussionTopic.ts:
```diff
@@ -0,0 +1,267 @@
+
+  if (startDateTime.getTime() === newestMeetingDate.getTime()) return
+
+  for (const interval of generateHourIntervals(startDateTime, newestMeetingDate)) {
```
+1 I think `generateHourIntervals` is being used to ensure a tolerable batch size. if that's correct, i don't think it's a good heuristic because meeting start times are not uniformly distributed. 

instead, in your query use `.orderBy('createdAt').limit(BATCH_SIZE)` and on each iteration of the loop set your `between` start time to the last time of the previous batch. be sure to support the case of 2 meetings starting at the exact same time!

jordanh replied:
Oh, nevermind. `summarySentAt` has no index. `createdAt it is.

mattkrick commented on packages/embedder/indexing/retrospectiveDiscussionTopic.ts:
```diff
@@ -0,0 +1,267 @@
+    // contain at least 1 discuss phase
+    const meetings = await r
+      .table('NewMeeting')
+      .between(interval[0], interval[1], {rightBound: 'closed', index: 'createdAt'})
```
-1 needs `orderBy`

mattkrick commented on packages/embedder/indexing/retrospectiveDiscussionTopic.ts:
```diff
@@ -0,0 +1,267 @@
+        r('meetingType')
+          .eq('retrospective')
+          .and(
+            r('summarySentAt').gt(0),
```
-1 `endedAt`, since summarySentAt could be wrong if the email didn't send

mattkrick commented on packages/embedder/indexing/retrospectiveDiscussionTopic.ts:
```diff
@@ -0,0 +1,267 @@
+  // Filter and sort comments as before
+  const filteredComments = comments
+    .filter((comment) => comment.threadParentId === parentId)
+    .sort((a, b) => a.threadSortOrder - b.threadSortOrder)
```
-1 `a.threadSortOrder < b.threadSortOrder ? -1 : 1`. shouldn't be an issue here, but it will better support NaN. back in the day safari would barf if you didn't write it like this

mattkrick commented on packages/embedder/indexing/retrospectiveDiscussionTopic.ts:
```diff
@@ -0,0 +1,267 @@
+    .table('RetroReflection')
+    .getAll(reflectionGroup.id, {index: 'reflectionGroupId'})
+    .run()
+  const promptIds = [...new Set(reflections.map((r) => r.promptId!))]
```
-1 `r.promptId` is guaranteed, no need for `!`
nit: even though it's in a different scope, calling things `r` in the presence of the rethinkdb `r` can be confusing

mattkrick commented on packages/embedder/indexing/retrospectiveDiscussionTopic.ts:
```diff
@@ -0,0 +1,267 @@
+      `\n`
+  for (const promptId of promptIds) {
+    if (!textForReranking) {
+      const prompt = await dataLoader.get('reflectPrompts').load(promptId)
```
-1 fetch all the prompts & then iterate over them so you only hit the DB once

mattkrick reviewed (changes requested):
didn't get to all of it, but here's the first pass

jordanh commented on packages/embedder/indexing/retrospectiveDiscussionTopic.ts:
```diff
@@ -0,0 +1,267 @@
+  markdown += `\n`
+
+  if (discussionSummary) {
+    markdown += `Further discussion was made. ` + ` ${discussionSummary}`
```
It's a good question. I'm not sure. Maybe?

The choice I made here was the default to the summary of the discussion if it exists in order to make this textual representation of a retrospective discussion a shorter token count. Using the summary almost certainly ensure that this text won't need to be sent to be summarized again before an embedding vector is calculated.

If we included the comments all the time – then we're maximizing the chance that rarer tokens might end up in the embed text and these tokens might affect the final embed vector in a useful way. However, we increase the odds that the embed text will need to be summarized before the vector is calculated.

I decided to "just be cheap" and try and minimize calls to the summarizer.

If we wanted to compare and contrast these approaches, we could always generate a second set of embed vectors `objectType: 'retrospectiveDiscussionNoSummary'` or something and do a bit of testing. Won't that be fun?!?

Your question makes me want to add a comment tho. I will do that.

jordanh commented on .env.example:
```diff
@@ -8,6 +8,11 @@ SERVER_ID='1'
 SOCKET_PORT='3001'
 
+# AI MODELS
+AI_EMBEDDING_MODELS='[{"model": "text-embeddings-inference:llmrails/ember-v1", "url": "http://localhost:3040/"}]'
```
I'd like to support having multiple configured embedding models for now. I'm going to keep this as it is, and we can revisit

mattkrick commented on packages/embedder/indexing/retrospectiveDiscussionTopic.ts:
```diff
@@ -0,0 +1,267 @@
+
+const pg = getKysely()
+
+export async function refreshEmbeddingsMeta(dataLoader: DataLoaderWorker) {
```
-1 rename to refreshRetroDiscussionTopicsMeta

mattkrick commented on packages/embedder/ai_models/TextGenerationInterface.ts:
```diff
@@ -0,0 +1,80 @@
+
+    try {
+      // console.log(`TextGenerationInterface.summarize(): summarizing from ${this.url}/generate`)
+      const res = await fetch(`${this.url}/generate`, {
```
+1 should we handle backpressure here? by default, the LLM server handles 128 concurrent requests. it'd be nice for this manager to either try again on a 500 code and/or to keep a running total to make sure we don't overload it

jordanh replied:
Implemented!

mattkrick commented on packages/embedder/ai_models/TextGenerationInterface.ts:
```diff
@@ -0,0 +1,80 @@
+      inputs: prompt,
+      parameters: {
+        maxNewTokens: maxNewTokens,
+        temperature: temperature
```
+1 this endpoint also accepts a `truncate` param. If we use that, we don't have to do a word count heuristic

jordanh replied:
Oh let me look at that...

jordanh replied:
Implemented truncate on the endpoint, and on node-llama-cpp-text-generation-inference

mattkrick commented on packages/embedder/indexing/numberVectorToString.ts:
```diff
@@ -0,0 +1,6 @@
+function numberVectorToString(vector: number[]) {
+  const stringArray = vector.map((num) => num.toString())
```
+1 not necessary, we coerce to a string below

mattkrick commented on packages/embedder/embedder.ts:
```diff
@@ -0,0 +1,250 @@
+  ejqRows.forEach((item) => {
+    const {refUpdatedAt} = ejqHash[makeKey(item)]!
+    const score = new Date(refUpdatedAt).getTime()
+    redisClient.zadd('embedder:queue', score, item.id)
```
-1 Could we reduce complexity by using PG xor redis for the job queue instead of both?

jordanh replied:
Yes, we could. Can we do this in a separate PR?

jordanh replied:
Let's merge this back to the trunk but not deploy it. I'd some help implementing the PG poller. I think it'll be rad.

I'll make sure to create a little "pre-prod" milestone we can track with a few issues I'm collecting

jordanh replied:
Issue added to milestone. I think this is the one to do first.

mattkrick reviewed (commented):
looking good! all done

jordanh commented on packages/embedder/ai_models/ModelManager.ts:
```diff
@@ -0,0 +1,179 @@
+  generationModels: GenerationModelConfig[]
+}
+
+export enum EmbeddingsModelTypes {
```
On it!

jordanh replied:
Changed!

jordanh commented on packages/embedder/ai_models/TextEmbeddingsInterface.ts:
```diff
@@ -0,0 +1,84 @@
+  Ember_v1 = 'llmrails/ember-v1'
+}
+
+const modelSubTypeDefinitions: Record<ModelSubTypes, EmbeddingModelParams> = {
```
On it!

jordanh commented on packages/embedder/ai_models/TextGenerationInterface.ts:
```diff
@@ -0,0 +1,80 @@
+  Zephyr_7b = 'TheBloke/zephyr-7b-beta'
+}
+
+const modelSubTypeDefinitions: Record<ModelSubTypes, GenerationModelParams> = {
```
Note: I got rid of the enums here too

jordanh commented on packages/embedder/ai_models/abstractModel.ts:
```diff
@@ -0,0 +1,73 @@
+export interface ModelConfig {
+  model: string
+  url?: string
```
This is a good catch – this was an artifact of an earlier design I had been working from. I've changed it on my branch and will push the changes

jordanh commented on packages/embedder/ai_models/abstractModel.ts:
```diff
@@ -0,0 +1,73 @@
+export interface GenerationModelConfig extends ModelConfig {}
+
+export abstract class AbstractModel {
+  public readonly modelConfigString: string
```
I'll get rid of this public, it's silly to have it. As for naming, I'm open to a different hierarchy.

modelType is nice. modelId is nice. I can make those changes.

jordanh commented on packages/embedder/embedder.ts:
```diff
@@ -0,0 +1,250 @@
+
+const POLLING_PERIOD_SEC = 60 // How often do we try to grab the lock and re-index?
+const Q_MAX_LENGTH = 100 // How many EmbeddingIndex items do we batch in redis?
+const WORD_COUNT_TO_TOKEN_RATIO = 3.0 / 2 // We multiple the word count by this to estimate token count
```
Oh wow. Color me impressed you have written exactly what we'd need to do this in JS. Ok. I am going to save this bit to the end to implement. Easy enough to add this to text-generation-interface and have the tokenizer function defined on a per modelId basis. Do you agree with that approach?

jordanh replied:
@mattkrick turns out the `/tokenize` input was just added 3 weeks ago! Checking now to see if it's on a container build yet...

https://github.com/huggingface/text-embeddings-inference/pull/139

jordanh replied:
I am going to punt this until after merge – here's what I think we should do:

* Create functions to tokenize ourselves and use those functions until...
* The `text-embedderings-inference`'s `/tokenize` endpoint is released to dockerhub, then update to use it

jordanh commented on packages/embedder/embedder.ts:
```diff
@@ -0,0 +1,250 @@
+})
+tracer.use('pg')
+
+export const pg = getKysely()
```
TIL - thx. Will fix

jordanh commented on packages/embedder/embedder.ts:
```diff
@@ -0,0 +1,250 @@
+tracer.use('pg')
+
+export const pg = getKysely()
+const dataLoader = getDataLoader() as DataLoaderWorker
```
omg 😊 thx for the catch

jordanh commented on packages/embedder/embedder.ts:
```diff
@@ -0,0 +1,250 @@
+  // that haven't been calculated nor exist in the EmbeddingsJobQueue yet
+  //
+  // Notes:
+  //   * `em.models @> ARRAY[v.model]` is an indexed query
```
I've tested this on a few hundred thousand rows on my system and EXPLAINing the queries it always looked to do the right thing... but let's keep an eye on it

mattkrick replied:
this query runs a _little_ slower than just doing a join, but not enough to care. That said, if denormalizing our data isn't getting us a speed boost, why not get rid of `models`? It introduces a 2nd source of truth & makes queries more complex to write. Check out the queries used for benchmarking

CURRENT
```sql
select "em".*, "model" from "EmbeddingsMetadata" as "em" 
left join lateral unnest(ARRAY[['Embeddings_ember_1']]::varchar[]) as "model" on true  
where ((not "em"."models" <@ ARRAY["model"]::varchar[] or "em"."models" is null)
and not exists (select "ejq"."id" from "EmbeddingsJobQueue" as "ejq" 
 where "em"."objectType" = "ejq"."objectType" 
 and "em"."refId" = "ejq"."refId" and "ejq"."model" = "model"))
```

PROPOSED
```sql
select * from "EmbeddingsMetadata" as em
left join "Embeddings_ember_1" as ember on ember."embeddingsMetadataId" = em.id
left join "EmbeddingsJobQueue" as ejq on ejq."embeddingsMetadataId" = em.id
where ember.id is null and ejq.id is null
```

jordanh replied:
I have a tendency to agree with you here. I suppose since we're queuing jobs one model at a time anyway, it's effort equivalent to just:

```
   FOR MODEL_TABLE_NAME IN MODEL_TABLE_NAMES:
      select * from "EmbeddingsMetadata" as em
      left join MODEL_TABLE_NAME as mt on mt."embeddingsMetadataId" = em.id
      left join "EmbeddingsJobQueue" as ejq on ejq."embeddingsMetadataId" = em.id
      where mt.id is null and ejq.id is null
```

Let's do this as a separate PR just to keep things need. I'll open the issue.

jordanh replied:
Linking this here: https://github.com/ParabolInc/parabol/issues/9449

jordanh commented on packages/embedder/embedder.ts:
```diff
@@ -0,0 +1,250 @@
+      refUpdatedAt: Date
+    }
+  } = {}
+  const makeKey = (item: {objectType: string | null; refId: string | null}) =>
```
Safety first! Fixed on my branch.

jordanh commented on packages/embedder/embedder.ts:
```diff
@@ -0,0 +1,250 @@
+      objectType: item.objectType,
+      refId: item.refId as string,
+      model: item.model,
+      state: 'queued' as DBInsert['EmbeddingsJobQueue']['state']
```
Fixed

jordanh commented on packages/embedder/embedder.ts:
```diff
@@ -0,0 +1,250 @@
+    let fullText = metadata?.embedText
+    try {
+      if (!fullText) {
+        fullText = await createEmbeddingTextFrom(jobQueueItem, dataLoader)
```
> +1 I think we're doing this after filtering orgs by feature flag to minimize workload. 

What I was sketching out the architecture, I was having a difficult time making the discovery of items to add to the metadata table stateless. That created a separation between "indexing items to metadata" from "creating embeddings". I'm open to figuring out ways to make this process simpler, better as we evolve it.

> Alternatively, we can consider embedText to be a function of the model because the input token limit dictates what the text is. If we do this, then we can remove embedText from the metadata & just store it on the model table.

I thought it might be handy to keep a "maximum tokens" version of the embedText on the metadata table – that way when we add additional models (which feels likely to me), we don't need to re-generate the embedText

jordanh commented on packages/embedder/embedder.ts:
```diff
@@ -0,0 +1,250 @@
+      // we're using word count as an appoximation of tokens
+      if (wordCount * WORD_COUNT_TO_TOKEN_RATIO > maxInputTokens) {
+        try {
+          const generator = modelManager.getFirstGenerator()
```
Oh you are positively correct. This was something I missed when I refactored to the new job queue. Great catch. I will fix this

jordanh commented on packages/embedder/ai_models/ModelManager.ts:
```diff
@@ -0,0 +1,179 @@
+    this.isValidConfig(config)
+    // Initialize embeddings models
+    this.embeddingModels = []
+    config.embeddingModels.forEach(async (modelConfig) => {
```
Updated!

jordanh commented on packages/embedder/ai_models/ModelManager.ts:
```diff
@@ -0,0 +1,179 @@
+  }
+
+  async maybeCreateTables(pg: Kysely<any>) {
+    for (const embeddingsModel of this.getEmbeddingModelsIter()) {
```
Good suggestion! Fixed.

jordanh commented on packages/embedder/ai_models/ModelManager.ts:
```diff
@@ -0,0 +1,179 @@
+  async maybeCreateTables(pg: Kysely<any>) {
+    for (const embeddingsModel of this.getEmbeddingModelsIter()) {
+      const tableName = embeddingsModel.getTableName()
+      const hasTable = await (async () => {
```
Removed!

jordanh commented on packages/embedder/indexing/retrospectiveDiscussionTopic.ts:
```diff
@@ -0,0 +1,267 @@
+  const discussPhase = newMeeting.phases.find((phase) => phase.phaseType === 'discuss')
+  const orgId = (await dataLoader.get('teams').load(newMeeting.teamId))?.orgId
+  if (orgId && discussPhase && discussPhase.stages) {
+    const metadataRows = discussPhase.stages.map(async (stage) => ({
```
Huh. Good catch on cruft. I must have refactored this at some point and did a poor job removing some cruft.

jordanh commented on packages/embedder/indexing/retrospectiveDiscussionTopic.ts:
```diff
@@ -0,0 +1,267 @@
+}
+
+export const upsertRetroDiscussionTopicFromNewMeeting = async (
+  newMeeting: MeetingRetrospective,
```
Well what do you know? Huh. Forget why I even had there in there. Nice catch.

jordanh commented on packages/embedder/ai_models/abstractModel.ts:
```diff
@@ -0,0 +1,73 @@
+export interface EmbeddingModelParams {
+  embeddingDimensions: number
+  maxInputTokens: number
+  tableSuffix: string
```
Hmm... I'm not sure where to put this... it'll vary with the `modelId`. Please have another look when I've pushed my 1st revisions

jordanh commented on packages/embedder/ai_models/ModelManager.ts:
```diff
@@ -0,0 +1,179 @@
+      DO $$
+      BEGIN
+      CREATE TABLE IF NOT EXISTS ${sql.id(tableName)} (
+        "id" SERIAL PRIMARY KEY,
```
TIL!

jordanh commented on packages/embedder/ai_models/ModelManager.ts:
```diff
@@ -0,0 +1,179 @@
+  try {
+    config.generationModels = AI_GENERATION_MODELS && JSON.parse(AI_GENERATION_MODELS)
+  } catch (e) {
+    throw new Error(`Invalid AI_EMBEDDING_MODELS .env JSON: ${e}`)
```
Nice catch! Me cut and paste? Never!

jordanh commented on packages/embedder/indexing/embeddingsTablesOps.ts:
```diff
@@ -0,0 +1,180 @@
+    .selectFrom('EmbeddingsMetadata as em')
+    .selectAll()
+    .leftJoin('EmbeddingsJobQueue as ejq', (join) =>
+      join.onRef('em.objectType', '=', 'ejq.objectType').onRef('em.refId', '=', 'ejq.refId')
```
Interesting thought. I think I'll keep it "fully denormalized" for now...let's see if it bugs us

jordanh commented on packages/embedder/embedder.ts:
```diff
@@ -0,0 +1,250 @@
+        try {
+          const generator = modelManager.getFirstGenerator()
+          if (!generator) throw new Error(`Generator unavailable`)
+          const textToSummarize = fullText.slice(0, generator.getModelParams().maxInputTokens)
```
This is a hack.

I am "in hate" with the state of tokenization. The fullText to embed is too long to be embedded directly, so we want to summarize it first. The GPT also has a maximum context window length. It will throw an error if you give it too much text. So, we need to make it shorter. Ideally the API would allow us to tokenize text in order to know how long it is. That API doesn't exist, so we just hack it down to a value we'll know fits before summarizing it.

I added a comment.

jordanh commented on packages/embedder/indexing/embeddingsTablesOps.ts:
```diff
@@ -0,0 +1,180 @@
+)`
+    }
+
+    if (metadata?.embedText !== fullText) {
```
The fullText parameter is always the fullText (it's never shorter), so we'll never lose data. I like the idea of renaming `EmbeddingsMetadata.embedText` to `EmbeddingsMetadata.fullText`. I will make that change.

jordanh commented on packages/embedder/embedder.ts:
```diff
@@ -0,0 +1,250 @@
+  // get the highest priority item and embed it
+  await dequeueAndEmbedUntilEmpty(modelManager)
+
+  setTimeout(() => tick(modelManager), POLLING_PERIOD_SEC * 1000)
```
This isn't what I want tho. If I use `setInterval` I could end up with more than one tick() running concurrently.

jordanh commented on packages/embedder/indexing/retrospectiveDiscussionTopic.ts:
```diff
@@ -0,0 +1,267 @@
+    .table('RetroReflectionGroup')
+    .get(discussStage.reflectionGroupId)
+    .run()
+  const reflections = await r
```
I made a few suggested changes to this function, but left the queries outside of Promise.all(); when running on our old data set, I've seen a few failures from mangled data and it's nice to have discretized errors thrown

jordanh commented on packages/embedder/indexing/retrospectiveDiscussionTopic.ts:
```diff
@@ -0,0 +1,267 @@
+  return markdown
+}
+
+export const createText = async (
```
Probably... we could either make 'objectType' have an implicit version or add a version field. I could be convinced either way

jordanh commented on packages/embedder/ai_models/TextGenerationInference.ts:
```diff
@@ -0,0 +1,88 @@
+import {truncate} from 'fs'
```
Note: this file was renamed from TextGenerationInterface to TextGenerationInference; for some reason when I wrote this over xmas I kept read "Inference" as "Interface". I've also renamed this repo: https://github.com/ParabolInc/node-llama-cpp-text-generation-inference

jordanh commented on packages/embedder/ai_models/helpers/fetchWithRetry.ts:
```diff
@@ -0,0 +1,61 @@
+import fetch, {RequestInfo, RequestInit, Response} from 'node-fetch'
```
I wrote this up real quick to retry in a mechanism similar to how we do API retries with Jira. Note that this supports both retrying at a specific time if `Retry-After` is returned by the server, or falling back to using an exponential backoff.

In either case, the caller must specify a deadline by which the API call must complete

jordanh replied:
Doh! Right! Thank you

jordanh commented:
Ok, @mattkrick – I think I addressed most of your review comments. I hope I did an ok job 🤞 

The majority I fixed right here on this branch. For a few items I created [a milestone](https://github.com/ParabolInc/parabol/milestone/193) and would love to involve others in shaping this up before using it with a general audience.

I'd like to get this merged (as it won't affect any releases until the service is enabled and deployed) and then work as a team to get the following implemented:

- #9432 
- #9439 
- #9433 
- #9436 
- #9437 
- #9438 

Please comment/edit the above issues with your input and/or add additional to the milestone, too :)

Could you take another pass on this PR?

jordanh commented on packages/embedder/ai_models/AbstractModel.ts:
```diff
@@ -0,0 +1,71 @@
+
+export abstract class AbstractModel {
+  public readonly url?: string
+  public modelInstance: any
```
Nuked

jordanh commented on packages/embedder/ai_models/AbstractModel.ts:
```diff
@@ -0,0 +1,71 @@
+  readonly tableName: string
+  constructor(config: EmbeddingModelConfig) {
+    super(config)
+    this.modelParams = this.constructModelParams(config)
```
Oh I see what you're saying now; it's about ergonomics. I can make that change real quick, sure

jordanh commented on packages/embedder/ai_models/AbstractModel.ts:
```diff
@@ -0,0 +1,71 @@
+export interface ModelConfig {
+  model: string
+  url?: string
```
Doh. Yes. For sure. That check was already there, just the type definition was wrong. This goes allllll the way back to when I first had tried to put node-llama-cpp directly in the embedder service (no URL if it's yourself). That's no longer the case anymore.

jordanh commented on packages/embedder/ai_models/helpers/fetchWithRetry.ts:
```diff
@@ -0,0 +1,61 @@
+  const timeoutId = setTimeout(() => controller.abort(), timeout)
+
+  try {
+    while (true) {
```
Sure thing, I'll do the while (now < deadline)

jordanh commented on packages/embedder/ai_models/helpers/fetchWithRetry.ts:
```diff
@@ -0,0 +1,61 @@
+
+      const retryAfter = response.headers.get('Retry-After')
+      // if Retry-After specified, use it; else fallback to exponential backoff
+      let waitTime = retryAfter ? parseInt(retryAfter, 10) * 1000 : Math.pow(2, attempt) * 1000
```
Yeah, you're right. That's messy. No use having the system hanging onto contexts it won't use. I'll fix this up.

jordanh commented on packages/embedder/embedder.ts:
```diff
@@ -0,0 +1,256 @@
+})
+tracer.use('pg')
+
+const getRedisClient = () => new RedisInstance(`embedder-${SERVER_ID}`)
```
Holy moses. Nice catch

jordanh commented on packages/embedder/embedder.ts:
```diff
@@ -0,0 +1,256 @@
+const getRedisClient = () => new RedisInstance(`embedder-${SERVER_ID}`)
+
+const refreshMetadata = async () => {
+  const dataLoader = getDataLoader() as DataLoaderWorker
```
Oh is that how that works? I'm a real dataloader noob. I had made an assumption that this was just retuning a singleton instance. Let me fix this up

mattkrick commented:
just 1 little thing & a merge conflict. merge when ready!
